protocol/packages/currencies/build/convert_case.rs

```diff
@@ -0,0 +1,31 @@
+use std::iter;
+
+pub(super) fn snake_case_to_upper_camel_case(mut input: &str) -> String {
+    iter::from_fn(move || {
+        input
+            .find('_')
+            .or_else(|| (!input.is_empty()).then_some(input.len()))
+            .map(|index| {
+                let substring = &input[..index];
+
+                input = input.get(index + 1..).unwrap_or("");
+
+                substring
+            })
+    })
+    .flat_map(|substring| {
+        let mut chars = substring.chars();
+
+        chars
+            .next()
+            .map(|first_character| first_character.to_ascii_uppercase())
+            .into_iter()
+            .chain(chars.map(|ch| ch.to_ascii_lowercase()))
```

Comment:
why introduce a lambda? would not `char::to_ascii_lowercase` work for you?

Reply:
I would have if I could, as you know I usually am the one to try simplify those. It takes a reference to `self`, instead of by value.